File: go32/iss_tim/pas/ISS_StartTimer.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

Function ISS_StartTimer(NewTProc : Pointer; NewTSpeed : DWord) : Boolean;
Var Counter : Word;
    TNumber : Word;
    NewTimerHandler: TSegInfo;
Begin
 Counter:=0; TNumber:=0;
 Repeat
  Inc(Counter);
  If Not ISS_TimersData[Counter].TActive Then TNumber:=Counter;
 Until (TNumber<>0) Or (Counter=ISS_MaxTimers);
 If TNumber=0 Then Begin
   ISS_TimerError:=ISS_TENoFree;
   ISS_StartTimer:=False;
   Exit;
  End;
 If Not NewIRQActive Then Begin
   Lock_Data(ISS_TimersData,SizeOf(ISS_TimersData));
   Lock_Data(DosMemSelector,SizeOf(DosMemSelector));
   Lock_Code(@SysTimerIRQ,DWord(@SysTimerIRQ_Dummy)-DWord(@SysTimerIRQ));
   Lock_Code(@UpdateUserTimers,DWord(@UpdateUserTimers_Dummy)-DWord(@UpdateUserTimers));
   NewTimerHandler.Offset:=@SysTimerIRQ;
   NewTimerHandler.Segment:=Get_CS;
   Get_PM_Interrupt(PIT_IRQ,OldTimer);
   Set_PM_Interrupt(PIT_IRQ,NewTimerHandler);
  End;
 ISS_TimersData[TNumber].TSpeed:=NewTSpeed;
 ISS_TimersData[TNumber].TCount:=0;
 ISS_TimersData[TNumber].TProc:=NewTProc;
 ISS_TimersData[TNumber].TActive:=True;
 SetTimerSpeed(GetTimerSpeed);
 ISS_StartTimer:=True;
End;

File: go32/iss_tim/ISS_StartTimer.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

bool ISS_StartTimer (void (*const proc) (void), uint32_t speed) {
  int i;

  for (i = 0; i < ISS_MaxTimers; i++)
    if (!ISS_TimersData[i].TActive) {
      if (!NewIRQActive) {
        __dpmi_paddr p;

        lock_SysTimerIRQ (true);
        __dpmi_get_protected_mode_interrupt_vector (PIT_IRQ, &OldTimer);
        p.offset32 = (uint32_t) SysTimerIRQ;
        p.selector = _go32_my_cs ();
        __dpmi_set_protected_mode_interrupt_vector (PIT_IRQ, &p);
      }

      ISS_TimersData[i].TSpeed = speed;
      ISS_TimersData[i].TCount = 0;
      ISS_TimersData[i].TProc = proc;
      ISS_TimersData[i].TActive = true;
      SetTimerSpeed (GetTimerSpeed ());
      return true;
    }

  ISS_TimerError = ISS_TENoFree;
  return false;
}
